src/network/rpc/client.rs

```diff
@@ -1,612 +1,289 @@
-		if let Some(sender) = self.response_sender.take() {
-			_ = sender.send(Err(error));
-		}
-	}
+	subxt_client: Arc<RwLock<avail::Client>>,
+	state: Arc<Mutex<State>>,
+	nodes: Nodes,
+	retry_config: RetryConfig,
 }
 
-struct GetSystemVersion {
-	response_sender: Option<oneshot::Sender<Result<String>>>,
-}
-
-#[async_trait]
-impl Command for GetSystemVersion {
-	async fn run(&mut self, client: &avail_subxt::avail::Client) -> Result<()> {
-		let res = client.rpc().system_version().await?;
+impl Client {
+	pub async fn new(
+		state: Arc<Mutex<State>>,
+		nodes: Nodes,
+		expected_genesis_hash: &str,
+		retry_config: RetryConfig,
+	) -> Result<Self> {
+		// try and connect appropriate Node from the provided list
+		// will do retries with the provided Retry Config
+		let (client, node, _) = Retry::spawn(retry_config.clone(), || async {
+			Self::connect_nodes_until_success(
+				nodes.shuffle(Default::default()),
+				ExpectedNodeVariant::new(),
+				expected_genesis_hash,
+				|_| futures::future::ok(()),
+			)
+			.await
+		})
+		.await?;
 
-		// send result back
-		// TODO: consider what to do if this results with None
-		if let Some(sender) = self.response_sender.take() {
-			_ = sender.send(Ok(res));
-		}
-		Ok(())
-	}
+		// update application wide State with the newly connected Node
+		state.lock().unwrap().connected_node = node;
 
-	fn abort(&mut self, error: Report) {
-		// TODO: consider what to do if this results with None
-		// TODO: maybe wrap error with specific error message per Command type implementation
-		if let Some(sender) = self.response_sender.take() {
-			_ = sender.send(Err(error));
-		}
+		Ok(Self {
+			subxt_client: Arc::new(RwLock::new(client)),
+			state,
+			nodes,
+			retry_config,
+		})
 	}
-}
-
-struct RequestRuntimeVersion {
-	response_sender: Option<oneshot::Sender<Result<RuntimeVersion>>>,
-}
 
-#[async_trait]
-impl Command for RequestRuntimeVersion {
-	async fn run(&mut self, client: &avail_subxt::avail::Client) -> Result<()> {
-		let res: RuntimeVersion = client
+	async fn create_subxt_client(
+		host: &str,
+		expected_node: ExpectedNodeVariant,
+		expected_genesis_hash: &str,
+	) -> Result<(avail::Client, Node)> {
+		let (client, _) = build_client(host, false).await.map_err(|e| eyre!(e))?;
+
+		// check genesis hash
+		let genesis_hash = client.genesis_hash();
+		info!("Genesis hash: {:?}", genesis_hash);
+		if let Some(cfg_genhash) = from_hex(expected_genesis_hash)
+			.ok()
+			.and_then(|e| TryInto::<[u8; 32]>::try_into(e).ok().map(H256::from))
+		{
+			if !genesis_hash.eq(&cfg_genhash) {
+				Err(eyre!(
+					"Genesis hash doesn't match the configured one! Change the config or the node url ({}).", host
+				))?
+			}
+		} else if expected_genesis_hash.starts_with(DEV_FLAG_GENHASH) {
+			warn!("Genesis hash configured for development ({}), skipping the genesis hash check entirely.", expected_genesis_hash);
+		} else {
+			Err(eyre!(
+				"Genesis hash invalid, badly configured or missing (\"{}\").",
```

Comment:
Same - we should add to log that we're skipping over to the next node (or exiting client).